Cloaking session secret for Flask

gridhead opened the pull request:
Cloaking session secret for Flask

Fixes #82.

Signed-off-by: Akashdeep Dhar <akashdeep.dhar@gmail.com>

gridhead committed:
Moved the session secret to __init__.py

Signed-off-by: Akashdeep Dhar <akashdeep.dhar@gmail.com>

t0xic0der commented:
This is still not safe. There needs to be a start script implemented to check if there are `.env` variables existing in the path or not, and if not so, they would create one and store the session secret on the first start. Subsequent uses would keep checking for the `.env` file and load up the session secret from it.

gridhead committed:
Made mandatory to load session secret from env file

If not available, session secrets are randomly generated and stored for further use

Signed-off-by: Akashdeep Dhar <akashdeep.dhar@gmail.com>

gridhead committed:
Added Python Dotenv as one of the dependencies

Minor version bump to v1.2.1-beta and removed __session__ variable from __init__.py

Signed-off-by: Akashdeep Dhar <akashdeep.dhar@gmail.com>

t0xic0der commented:
With changes made via https://github.com/t0xic0der/supervisor-frontend-service/pull/87/commits/7ca55d08400475ed436497fad32a661f822f417a and https://github.com/t0xic0der/supervisor-frontend-service/pull/87/commits/5b6a0c448c1725037f0c456f5e9c8b40c7d3f261, session secrets are randomly generated in the first-time execution and then reused thereafter. From a security standpoint, the presence of the `sesskeys` environment variable is first checked and if it is not found, only then are the new session secrets generated.

Generation of new session secrets would mean that the older existing sessions which made use of the previous session secret would no longer be valid. This is not a very big deal as the user would be shown the following `403 Forbidden` page - from where they can head over to the login page and log back in with the same credentials.

![image](https://user-images.githubusercontent.com/49605954/115979516-f09ceb00-a5a3-11eb-9fdf-9cb8042ffb00.png)

Do note that the session secrets are not regenerated on every execution as once a valid `.env` is found, the server would try to make use of that. Session secrets can be explicitly regenerated if there is a suspicion of session data being compromised by deleting the `.env` file inside the container (if using the Docker image) or from the install location (if using the COPR) and restarting the server. Once the server starts, it would not find a valid `.env` file as a result of the deletion and hence, regenerate a new one and store it.

The following is the prompt when the session secrets are not found and hence, generated and stored.

```
 * Setting up the session secret into environment variables...
 * SuperVisor Frontend Service v1.2.1-beta
 * Port number    : 9696
 * IP version     : 4
 * Logs state     : Errors only
 * Serving Flask app "main" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://0.0.0.0:9696/ (Press CTRL+C to quit)
```

The following is the prompt when the session secrets are found and hence, reused.

```
 * Loading up the session secret from environment variables...
 * SuperVisor Frontend Service v1.2.1-beta
 * Port number    : 9696
 * IP version     : 4
 * Logs state     : Errors only
 * Serving Flask app "main" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://0.0.0.0:9696/ (Press CTRL+C to quit)
```

mscherer commented on src/svfrontend/main.py:
```diff
@@ -42,7 +45,21 @@
+    click.echo(" * Loading up the session secret from environment variables...")
+else:
+    # If session secrets are not available in the environment variables, create and store them
+    randstrg = "".join(secrets.choice("ABCDEF" + "0123456789") for i in range(64))
```
I think the entropy is sufficient for that use case, but it seems that you can directly use os.random if you want to generate the key, and not be limited to 16 chars cf https://flask.palletsprojects.com/en/1.1.x/quickstart/#sessions 

Now, of course, I do not know if using really random data would work well with dotenv.

mscherer replied:
In fact, maybe secrets.token_hex would be a simpler choice (along comment explaining why token_hex and not token_bytes, cause token_bytes might generate data that dotenv might not handle well, like ' , various \, etc.

t0xic0der replied:
> I think the entropy is sufficient for that use case, but it seems that you can directly use os.random if you want to generate the key, and not be limited to 16 chars cf https://flask.palletsprojects.com/en/1.1.x/quickstart/#sessions
> 
> Now, of course, I do not know if using really random data would work well with dotenv.

The question is, how viable would that replacement be. The following is what I mean when I ask about viability.
1. Is there going to be a significant amount of gain in terms of security and/or performance point of view when `os.random` is used over `secrets`?
2. I have used to a whopping 64 char length here to greatly decrease the likeliness of it being discovered via brute force. Would changing that something else be the right thing to do?

> In fact, maybe secrets.token_hex would be a simpler choice (along comment explaining why token_hex and not token_bytes, cause token_bytes might generate data that dotenv might not handle well, like ' , various , etc.

A total +1 to this! It would avoid me to do all the monkey business I have done with explicitly including the hex characters and the code would end up looking far more elegant. :)

mscherer commented on src/svfrontend/main.py:
```diff
@@ -42,7 +45,21 @@
+else:
+    # If session secrets are not available in the environment variables, create and store them
+    randstrg = "".join(secrets.choice("ABCDEF" + "0123456789") for i in range(64))
+    with open(".env", "w") as envrfile:
```
I think this will work most of the time, but if the flask code is in /srv/ and I run it with my current directory in /tmp, dotenv will search in /srv (per my pre breakfast understanding of https://github.com/theskumar/python-dotenv/blob/master/src/dotenv/main.py#L252 ), not in /tmp. 

But this part of the code create it in the current directory.

I am bringing this up because systemd run service with working directory set as /, and if the application is install in the regular python path, it can be run from /, and it will create a /.env file

mscherer replied:
Also, it might be worth doing a pass at unix permission of the file. I do not know the default value for umask, but writing a world readable file with secret do not seems like a good practice. 

As far as I understand the code, the worst case scenario is that if someone can get the secret key, the attacker can create a signed cookie for flask and bypass authentication (since the authentication is verified by having a valid session ID).

t0xic0der replied:
> I think this will work most of the time, but if the flask code is in /srv/ and I run it with my current directory in /tmp, dotenv will search in /srv (per my pre breakfast understanding of https://github.com/theskumar/python-dotenv/blob/master/src/dotenv/main.py#L252 ), not in /tmp.

Good point. But, it is going to look for the `.env` file in the directory where the code is located but at the time of creating the environment variable, it is going to do that in the current working directory (wherever it is executed from) - which is definitely not something intended. A couple of hacks to find out the `pwd` first, and then explicitly storing the `.env` file there should fix this - what do you think?

> I am bringing this up because systemd run service with working directory set as /, and if the application is install in the regular python path, it can be run from /, and it will create a /.env file

But (please correct me if I am mistaken), we do have the ability to change the working directory for a systemd service using one of the directives, don't we?

> Also, it might be worth doing a pass at unix permission of the file. I do not know the default value for umask, but writing a world readable file with secret do not seems like a good practice.
>
> As far as I understand the code, the worst case scenario is that if someone can get the secret key, the attacker can create a signed cookie for flask and bypass authentication (since the authentication is verified by having a valid session ID).

+1 to this too! I would have to do some reading before I would be able to deliver a servicable fix for this. This should not a problem if users are running it as a container (unless they have a way to enter the containers and explore them) but is a threat if it is installed using the COPR method.

mscherer replied:
> But (please correct me if I am mistaken), we do have the ability to change the working directory for a systemd service using one of the directives, don't we?

I think that's what WorkingDir do, if I am not mistaken and if I understand the question.
